src/c++/library/common.cc

```diff
@@ -232,6 +236,113 @@ InferInput::SetBinaryData(const bool binary_data)
   return Error::Success;
 }
 
+Error
+InferInput::GetStringCount(size_t* str_cnt) const
+{
+  int64_t str_checked = 0;
+  size_t remaining_str_size = 0;
+
+  size_t next_buf_idx = 0;
+  const size_t buf_cnt = bufs_.size();
+
+  const uint8_t* buf = nullptr;
+  size_t remaining_buf_size = 0;
+
+  // Validate elements until all buffers have been fully processed.
```

Comment:
Don't we do the same check in core?

Reply:
Yes, but we are also bringing the checks to the client side.

Reply:
What's the value in doing the check on client first if it's going to be checked a second time in core anyways? 


For the case where size is wrong, we will catch it early and not send to core, which could be seen as a benefit.

However, I think the assumption here is that this is the bad/unexpected case, and for users who are correctly setting shapes/sizes, they will incur the cost of making the same check twice on each inference unnecessarily, right?

Reply:
True. At the end it's up to how much overheads it adds to the client. For string input in this case, let's say there are N strings and each string has a size M. The overhead added for checking number of strings is O(N). It's negligible comparing copying strings which is O(M*N).

Reply:
CC @tanmayv25 @GuanLuo if you have any thoughts

Reply:
Will defer to Tanmay as he has more context on the original ask.

Personally I agree with Ryan on it being duplicating check on the inference process. I see value in having such a check in client for development / debug purpose, i.e. not requiring an running server for catching this type of misuse, and in such case the validation path should be controlled by toggling debug setting

Reply:
Alternative to this iterative checks is directly counting strings in function ```AppendFromString ```. But ```AppendFromString``` internally calls ```AppendRaw``` so we need to disable the string input shape checks if ```AppendRaw``` is called, and this method does not work on shared memory.

Reply:
I understand the added overhead with testing for string tensors on the client side. The reason is that we need to iterate through data to correctly determine the size of tensor being sent to the server. There is a performance impact. 

The reason for adding the bytesize checks in the client was to prevent client from adding unnecessary load to the server and prevent unnecessary data processing in the server frontends.

I propose a middle ground. How about just checking the shape of the tensors whose datatype is of fixed byte size? 
It should cover most of the cases by not allowing the client library from sending extra bytes. For TYPE_BYTES, we can rely on the Core side check to come up with an error. 
@rmccorm4 @GuanLuo @yinggeh thoughts?

Reply:
> I propose a middle ground. How about just checking the shape of the tensors whose datatype is of fixed byte size?
It should cover most of the cases by not allowing the client library from sending extra bytes. For TYPE_BYTES, we can rely on the Core side check to come up with an error.

That works for me :+1: 

If in the future we decide to bring back the TYPE_BYTES check on client-side for C++ client, I think this should involve some sanity performance testing, as PA will use the C++ client under the hood, and may bottleneck PA's ability to send requests at a high rate.

Reply:
Sounds good